### ISSUE TITLE
perf: performance issue when there is only barrier passing the cluster

### ISSUE THREAD

**Issue by st1page** [risingwavelabs/risingwave#17646]
In some real-world scenarios, we have encountered this situation.
- No source throughput and other data, only barrier in the stream
- barrier number pie up, but max commited epoch keep increasing
- No very slow tasks in the await tree
- high CPU usage, hyper(httpserver) and tokio in the flamegraph

In this situation, the only flow in the graph is the barrier. In our assumption, when there is no data between barriers, the barriers should flow through quickly. Otherwise, if the time for barriers to pass through the entire graph exceeds one second, then barriers will continue to accumulate, and the cluster will always be in a state of backpressure. more infos here https://www.notion.so/risingwave-labs/CVTE-2024-07-10-barrier-only-ecc23aa5b9ee4664a97a17c97a25d709?pvs=4

Below are some suspected causes and potential optimizations we are currently considering for this situation:
- Hyper + Tokio's performance issues when handling small network packets. We have discovered that in this scenario, the CPU usage is abnormally high, and there are many atomic variable and lock calls from Hyper in the flame graph.
- backpressure with limiting concurrent barrier number in exchange. We introduce exchange_concurrent_barriers to limit concurrent barriers in exchange in https://github.com/risingwavelabs/risingwave/pull/9427, which makes the backpressure more sensitive. When there are only barriers on the stream, the only barrier stream will be slow when exchange_concurrent_barriers = 1 and in that case, all barrier messages will pass the streaming message sequentially, in other words, the upstream must send the barrier until the downstream actor has consumed that message and send back a ack message.
    - When sending barriers to the downstream system, it is necessary to wait for the downstream to receive and consume the barrier before returning the RPC permit to the upstream. Only then can the upstream send the next barrier. 
    - Even after adjusting the value of exchange_concurrent_barriers to a larger number, the problem was not resolved. This suggests that the issue lies in another direction, where there is no batching occurring. That is, for every barrier consumed by the downstream, an RPC is sent to the upstream to return the permit. In other words, there is no batching of permits.
- The streaming pipeline is quite extensive. The triggering scenario involves a left-deep tree join of 20 tables, which results in the barrier flow having to pass through dozens of exchanges along the entire pipeline.
- Barrier interval. Obviously, increasing the barrier interval can resolve this situation. On the contrary, we could also use a smaller barrier interval to reproduce the issue easier.

Even with some conjectures, I believe we first need to find a way to reliably reproduce this situation before attempting to make improvements. This way, we can verify the effectiveness of the optimizations.

- [ ] reproduce the case and invesitigate

**Comment by BugenZhao** [comment]
Will it be caused by the problem described in https://github.com/risingwavelabs/risingwave/pull/17612?

**Comment by fuyufjh** [comment]
Can't reproduce

**Comment by fuyufjh** [comment]
Recurs today in another case

**Comment by hzxa21** [comment]
Recurs today in another case

**Comment by fuyufjh** [comment]
The overhead (CPU usage) comes from the passing the barriers. We have to admit that passing a barrier will cost some overhead, although the overhead is neglectable in most cases.

It reminds me of #6943, where we set the `in_flight_barrier_nums` to 10000 to eliminate the upper bound. This also causes a side effect - the overhead of passing barriers grows along with the number of in-flight barriers.

As written in #6943, the motivation was because of OOM issues. Shall we reduce number from 10000 to some lower number such as 1000 or 100, in order to reach a balance point between these two issues?

**Comment by hzxa21** [comment]
It occurred again in a user cluster (4x48C CN) when I lower their barrier interval from 10s to 1s. It is running v2.0.2. I managed to take cpu flamegraph when the issue happens:
![1](https://github.com/user-attachments/assets/8f33f132-13ca-4788-82dc-87adf7445789)
![2](https://github.com/user-attachments/assets/f637fdea-949b-4032-b5f1-1a0db2359546)
![3](https://github.com/user-attachments/assets/53d19092-3fbe-4a8b-b804-3fff729b257a)
![4](https://github.com/user-attachments/assets/54cfc5db-bdd9-425d-80af-5a07e3680f3a)

**Comment by hzxa21** [comment]
Found one weird phenomenon from the occurrence above:
- With barrier_interval_ms = 20000, source_rate_limit = 20, we experience ~100% CPU usage and increasing barrier latency with no source tp and actor tp. Triggering manual recovery doesn't help.
- With barrier_interval_ms = 20000, source_rate_limit = 50, CPU is reduced from ~100% to ~30-40% and barrier latency is stable with constant source tp and actor tp.

That means the issue is resolved if there are data between barriers with no change in barrier interval changes.

**Comment by BugenZhao** [comment]
Find a way to reproduce "only barrier passing":

Launch RisingWave with

```
RUST_LOG=events::stream::message::chunk=DEBUG \
RW_BARRIER_INTERVAL_MS=1 \
RW_PARALLELISM=256 \
risingwave play
```

then

```sql
SET streaming_parallelism TO 128;

CREATE SOURCE a (v1 int)
  WITH (
    connector = 'datagen',
    fields.v1.kind = 'sequence',
    fields.v1.start = '1',
    datagen.rows.per.second = '1'
  )
  FORMAT PLAIN
  ENCODE JSON;

CREATE MATERIALIZED VIEW mv
AS
SELECT
max(v1)
FROM a
GROUP BY v1;
```

will find that CPU is fully utilized but no chunk messages pass between actors, as no `events` logs printed and selecting from `mv` returns empty set.

By reducing `streaming_parallelism` to a lower value, we may observe that data is correctly generated and processed.

**Comment by hzxa21** [comment]
Can we reproduce the issue if we replace the datagen source with an empty table with no connector?

**Comment by BugenZhao** [comment]
> Can we reproduce the issue if we replace the datagen source with an empty table with no connector?

Yes. It turns out that a 128x128 hash-shuffle is too busy while aligning the barrier.

**Comment by hzxa21** [comment]
May worthing taking a cpu flamegraph in your repro to see whether it is similar to the production flamegraph above.

**Comment by MrCroxx** [comment]
Hi, folks. May I ask if all of the cases above were configured with rate limiters?

**Comment by hzxa21** [comment]
Optimization merged in 2.3
